lib/bitrange.h

```diff
@@ -51,4 +58,714 @@ class bitranges {
     iter end() const { return iter(bits.end()); }
 };
 
+class JSONGenerator;
+class JSONLoader;
+
+namespace BitRange {
+
+namespace Detail {
+/**
+ * @return the result of dividing @dividend by @divisor, rounded towards
+ * negative infinity. This is different than normal C++ integer division, which
+ * rounds towards zero. For example, `-7 / 8 == 0`, but
+ * `divideFloor(-7, 8) == -1`.
+ */
+inline int divideFloor(int dividend, int divisor) {
+    const int quotient = dividend / divisor;
+    const int remainder = dividend % divisor;
+    if ((remainder != 0) && ((remainder < 0) != (divisor < 0))) return quotient - 1;
+    return quotient;
```

Comment:
Does this get properly folded to a shift when divisor is a constant power of 2?  Similarly, moduloFloor should get folded to an and

Reply:
Testing in the [Compiler Explorer](https://godbolt.org/z/Ks3rf7e3T):
* clang: folds it to a right shift
* gcc: does _not_ fold to a right-shift
(I tested several different -O levels with gcc, but didn't get it to fold.)

Any suggestions how to restructure the code to get gcc to fold it to a shift?

Reply:
And neither compiler does the fold to and for moduloFloor 🙁

Reply:
Updated Compiler Explorer with both divideFloor and moduloFloor: https://godbolt.org/z/8a1KeYGon

Reply:
I've got a solution to this, but I'm concerned about its portability. Will open a separate PR for that to get feedback after this has merged.